source/common/common/key_value_store_base.cc

```diff
@@ -73,21 +101,27 @@ void KeyValueStoreBase::addOrUpdate(absl::string_view key_view, absl::string_vie
     ASSERT(false);
     return;
   }
+  absl::optional<std::chrono::seconds> absolute_ttl = absl::nullopt;
+  if (ttl) {
+    absolute_ttl.emplace(ttl.value() + std::chrono::duration_cast<std::chrono::seconds>(
+                                           time_source_.monotonicTime().time_since_epoch()));
```

Comment:
I think Snow is right, since we're loading from disk after restarts, I don't think `std::chrono::steady_clock` would allow us to capture the time elapsed since the value was persisted.